lib/Sema/DerivedConformanceDifferentiable.cpp

```diff
@@ -841,28 +788,35 @@ static void checkAndDiagnoseImplicitNoDerivative(TypeChecker &TC,
     bool conformsToDifferentiable =
         TC.conformsToProtocol(varType, diffableProto, nominal,
                               None).hasValue();
-    bool isConstantProperty = vd->isLet() && vd->hasInitialValue();
     // If stored property should not be diagnosed, continue.
-    if (conformsToDifferentiable && !isConstantProperty)
+    if (conformsToDifferentiable && !vd->isLet())
       continue;
     // Otherwise, add an implicit `@noDerivative` attribute.
     vd->getAttrs().add(
         new (TC.Context) NoDerivativeAttr(/*Implicit*/ true));
     auto loc =
         vd->getLoc().isValid() ? vd->getLoc() : DC->getAsDecl()->getLoc();
     assert(loc.isValid() && "Expected valid source location");
-    // Diagnose stored property with fixit.
+    // If nominal type can conform to `AdditiveArithmetic`, suggest conforming
+    // adding a conformance to `AdditiveArithmetic`.
+    // `Differentiable` protocol requirements all have default implementations
+    // when `Self` conforms to `AdditiveArithmetic`, so `Differentiable`
+    // derived conformances will no longer be necessary.
+    bool nominalCanDeriveAdditiveArithmetic =
+        DerivedConformance::canDeriveAdditiveArithmetic(nominal, DC);
     if (!conformsToDifferentiable) {
       TC.diagnose(loc,
                   diag::differentiable_nondiff_type_implicit_noderivative_fixit,
-                  vd->getName())
+                  vd->getName(), nominal->getName(),
+                  nominalCanDeriveAdditiveArithmetic)
           .fixItInsert(vd->getAttributeInsertionLoc(/*forModifier*/ false),
```

Comment:
Note: the fixit location doesn't seem right. It should be before the `var/let` specifier, not after.
This is likely a bug in `ValueDecl::getAttributeInsertionLoc`.

Example:
```swift
struct Example: Differentiable {
    @usableFromInline let bad1: Float
    let bad2: Int
}
```
```
fixit.swift:2:27: warning: synthesis of the 'Differentiable.move(along:)' requirement for 'Example' requires all stored properties to be mutable; use 'var' instead, or add an explicit '@noDerivative' attribute, or conform 'Example' to 'AdditiveArithmetic'
    @usableFromInline let bad1: Float
                          ^
                      @noDerivative
fixit.swift:3:9: warning: stored property 'bad2' has no derivative because it does not conform to 'Differentiable'; add an explicit '@noDerivative' attribute, or conform 'Example' to 'AdditiveArithmetic'
    let bad2: Int
        ^
    @noDerivative
```

Reply:
Actually, `ValueDecl::getAttributeInsertionLoc` is fine, but wasn't used properly here.
Fixed in https://github.com/apple/swift/pull/25437/commits/8fc3cd148eb4f66f17bafff21bf28e0fff391567.

---

Before:
```
    let bad2: Int
        ^
    @noDerivative
```

After:
```
    let bad2: Int
    ^
    @noDerivative
```